kernel/src/oops.rs

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Kernel oops handling.
+//!
+//! In Asterinas, the Rust panic leads to a kernel oops. The kernel oops behave
+//! as resettable exceptional control flows. If the kernel oops happens too
+//! frequently, the kernel will panic and halt the system.
+//!
+//! Though we can recover from the Rust panics. It is generally not recommended
+//! to make oops as a general exception handling mechanism. Handle [`Result`]
+//! instead.
+
+use alloc::{
+    boxed::Box,
+    string::{String, ToString},
+};
+use core::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+
+use ostd::panic;
+
+static PANIC_ON_OOPS: AtomicBool = AtomicBool::new(true);
+
+/// Initializes the default panic-on-oops setting from the kernel command line.
+pub fn init_panic_on_oops() {
+    // TODO: Control the kernel commandline parsing from the kernel crate.
+    PANIC_ON_OOPS.store(true, Ordering::Relaxed);
+}
+
+/// Executes the given function and catches any panics that occur.
+///
+/// All the panics in the given function will be regarded as oops. If a oops
+/// happens, this function returns `None`. Otherwise, it returns the return
+/// value of the given function.
+pub fn catch_panics_as_oops<F, R>(f: F) -> Option<R>
+where
+    F: FnOnce() -> R,
+{
+    if PANIC_ON_OOPS.load(Ordering::Relaxed) {
+        return Some(f());
+    }
+
+    let result = panic::catch_unwind(f);
+
+    match result {
+        Ok(result) => Some(result),
+        Err(err) => {
+            let count = OOPS_COUNT.fetch_add(1, Ordering::Relaxed);
+
+            log::error!("Oops! {}", err.downcast::<String>().unwrap());
+
+            if count >= MAX_OOPS_COUNT {
+                // Too many oops. Panic the kernel.
+                //
+                // Note that for nested `catch_panics_as_oops` it still works as
+                // expected. The outer `catch_panics_as_oops` will catch the panic
+                // and found that the oops count is too high, then panic the kernel.
+                panic!("Too many oops. The kernel panics.");
+            }
+
+            None
+        }
+    }
+}
+
+/// The maximum number of oops allowed before the kernel panics.
+///
+/// It is the same as Linux's default value.
+const MAX_OOPS_COUNT: usize = 10_000;
+
+static OOPS_COUNT: AtomicUsize = AtomicUsize::new(0);
+
+#[ostd::panic_handler]
+fn panic_handler(info: &core::panic::PanicInfo) -> ! {
+    let message = info.message().to_string();
+
+    // Raise the panic and expect it to be caught.
+    panic::begin_panic(Box::new(message.clone()));
```

Comment:
Heap allocation inside the panic handler should be avoided to make the panic handler more robust. Even in face of Rust panics due to out-of-memory errors.

Reply:
It is a little bit hard to eliminate heap allocation for it. We would unwind the stack so nothing can be preserved on the stack. There must be some global storage for such panic information. Both the `unwinding` crate and the standard library uses `Box` so we need some extra efforts to put it somewhere else. I want to postpone it and added a TODO for it.